asserts: define type for revision related errors

emgee opened the pull request:
The error type can be used to improve handling when an assertion cannot be added because the revision is incorrect.

snappy-m-o commented:
Can one of the admins verify this patch?

snappy-m-o commented:
Can one of the admins verify this patch?

snappy-m-o commented:
Can one of the admins verify this patch?

pedronis commented on asserts/database.go:
```diff
@@ -89,6 +89,13 @@ var (
 )
 
+// InvalidRevisionError describes an error related to the revision of an assertion.
+type InvalidRevisionError string
```
ConflictingRevisionError? SupersededRevisionError?

given that we are introducing this, I would really make struct carrying the two relevant revno numbers (that we put in the error message)

pedronis commented:
made a comment about naming and type of the error,

could you merge master into this? I don't think there should be conflicts or issues but it's been a long time;

ideally the relevant REST API endpoint should return 409 when it sees this but I can take care of that later

emgee commented:
I renamed the error type. It now has a very specific purpose so went with SupercededRevisionError.

I also updated the mem backstore implementation, and added tests for both backstores included in the package. Something for later perhaps, but it might make sense to have a test suite that runs a set of common tests against a set of backstores to check they're in sync.

niemeyer commented on asserts/database.go:
```diff
@@ -89,6 +89,18 @@ var (
 	ErrNotFound = errors.New("assertion not found")
 )
 
+// SupersededRevisionError is the error returned when an assertion is added
```
This error type can be made more useful if we cook it with a wider scope:
```
// SupersededRevisionError is observed when an operation attempts to deal with an assertion
// whose revision is no longer current.
type SupersededRevisionError struct {
        superseded, current int
}

func (e *ObsoleteRevisionError) Error() string {
        return fmt.Sprintf("cannot handle assertion with superseded revision %d, current is %d", e.superseded, e.current)
}
```

I'd suggest making both fields private if possible, as suggested above, as that gives us room for tweaking the way we use it without breaking clients. For example, we might use this in cases we don't actually have all the details, or don't want to leak all the details.

niemeyer replied:
(renamed to "SupersededError", which is shorter and clear enough)

pedronis replied:
I think surfacing at current could be useful, is the bit of info the caller doesn't quite have without doing extra finds/gets (revision itself is something it should have at end)

emgee replied:
Shortening to SupersededError seems fine to me, but "no longer current" isn't quite accurate - the error is also returned when the currently stored revision is the *same* as the revision of the assertion being added.

Making Current public could be useful, but is not essential to the use-case I needed an error type for. If I knew the current revision was different to the revision being added I could skip one database lookup.

niemeyer replied:
That's interesting. I suggest handling those cases differently. Attempting to add an assertion to the database that matches the existing revision should be silently accepted if the assertion content matches exactly (it's a no-op, and the end result is the same, idempotent), but if the assertion doesn't match it should be a harsh failure rather than pretending the assertion is simply not up-to-date enough. Having two conflicting statements with the same revision seems like something is either seriously broken, or someone is trying to maliciously abuse the system. Both are cases worth raising prominently, rather than pretending it's just an old assertion we don't care about.

What do you think?

emgee replied:
Making Add idempotent for an identical, current assertion is the exact use case I have for this error type. :)

Moving the idempotent check into the asserts package is doable. It would obviously be a small change in behaviour, but I can't imagine anything would break. I think an idempotent check would be best handled by Database.Add so each backstore doesn't have to deal with the issue. We'd still need this error type, so I would suggest a new branch if we decide to go that way.

Two different assertions with the same revision is not necessarily as bad as you suggest. It could just be two concurrent processes trying to update an assertion ... just a conflict due to an optimistic lock style check really.

niemeyer replied:
Sorry, that's not appropriate for all cases. Some assertions will be built at the client side, and for those version skews are not so hard to observe.

At the same time, going back to your earlier point, if we can have a matching revision being an error, that's not quite a superseded revision. It's more like a conflict.

So, here is a suggested phrasing, making the revisions public too after the comments above, and going back to a general error type that can hopefully address all cases we'll need it for:
```
// RevisionError indicates a revision improperly used for an operation.
type RevisionError struct {
        Used, Current int
}

func (e *RevisionError) Error() string {
        if e.Used == 0 || e.Current == 0 {
                return fmt.Errorf("assertion revision %d is unknown", e.Used)
        }
        if e.Used == e.Current {
                return fmt.Sprintf("revision %d is already the current revision", e.Used)
        }
        if e.Used < e.Current {
                return fmt.Sprintf("revision %d is older than current revision %d", e.Used, e.Current)
        }
        return fmt.Sprintf("revision %d is more recent than current revision %d", e.Used, e.Current)
}

emgee replied:
Just implemented and pushed this, although it's not quite the same as proposed. Hopefully the test explains the reasons for that.

niemeyer commented:
LGTM with the details mentioned.

niemeyer commented:
Sorry for the back and forth, by the way. I see Samuele made a similar suggestion in the opposite direction, which sounded sensible too. My suggestion is mainly on the opposite direction so we don't over specialize it, cooking an error that is only useful for adds.

pedronis commented on asserts/database.go:
```diff
@@ -89,6 +89,27 @@ var (
+	if e.Used < 0 {
+		return fmt.Sprintf("used assertion revision %d is unknown", e.Used)
+	}
+	if e.Current < 0 {
```
I think it was deliberate to let one use Current == -1 to get  a message about Used being wrong/unknown

emgee replied:
But then {Used: 2, Current: -1} would produce a message of, "assertion revision 2 is unknown", that doesn't make sense.

We could drop the first word - used or current - from the message but with the invalid revision's value?

pedronis replied:
well I think the intention was to have indeed that sort of message

pedronis commented on asserts/database.go:
```diff
@@ -89,6 +89,24 @@ var (
+
+func (e *RevisionError) Error() string {
+	if e.Used < 0 || e.Current < 0 {
+		return fmt.Sprintf("assertion revision is unknown")
```
could add a TODO here that says something about the message may need tweaking once we hit a real use case for this one
